### PR TITLE
L9 pandas

### PR DESCRIPTION
**Pull request by Tasenga**
In 'analysis_of_user_actions' module the following data were output with using Jupyter notebook:

    - retrieving data that can't be processed;
    - percentage of operations (pie chart);
    - the total number of operations by weeks;
    - top 10 most lengthy searching operations indicating time spent;
    - histogram of input data for each operation;
    - graph of the number of successful, conditionally successful (warning) and unsuccessful results;
    - dependence of the time spent on the amount of input data.

### CODE REVIEW

**Comment by sleshJdev** [comment]
The PR relates to the lecture nine but its name is L10, why? It's confused a bit

**Review by sleshJdev — CHANGES_REQUESTED** [review]
Overall looks good, but I have a feeling that code is not quite optimized and we do a lot of work multiple times even though it can be done once when we read data. I mean this function:
```
def process_table(base, type_msg):
    drop = base[base['message'].apply(lambda x: not x.startswith(type_msg))].index
    new_table = base.drop(drop).loc[:,['actionId', 'createdDateTime', 'message']]
    new_table['action'] = new_table.apply(lambda x: parse_msg(x['message'])[1], axis=1)
    new_table['msg'] = new_table.apply(lambda x: parse_msg(x['message'])[2], axis=1)
    new_table.drop(columns='message', inplace=True)
    return new_table
```
We can derive required values in advance, e.g.(I do not pretend that this is the best solution, but an idea just that):
```
def parse_row(row: Series):
    state, user_action, _  = re.findall(r'\[(.*?)\]', row.message)
    return user_action, row.createdDateTime, state == 'START'
def collapse(group: DataFrame):
    start_action, started_at, is_start = parse_row(group.iloc[0])
    end_action, ended_at, _ = parse_row(group.iloc[1]) if group.index.size == 2 else (None, None, None)
    if not is_start: # just swap values
        start_action, end_action = end_action, start_action
        started_at, ended_at = ended_at, started_at
    return Series({
        'start_action': start_action,
        'started_at': started_at,
        'end_action': end_action,
        'ended_at': ended_at,
        'duration': ended_at - started_at if ended_at and started_at else None 
    })
user_actions=pd.read_csv('user_actions.csv', dtype={'message': str}, parse_dates=['createdDateTime'])\
    .groupby(['actionId', 'currentUser'])[['message', 'createdDateTime']]\
    .apply(collapse).reset_index()
user_actions.head()
```
It will produce you data frame like this:
| # | actionId                             | currentUser     | start_action | started_at          | end_action | ended_at            | duration | 
|---|--------------------------------------|-----------------|--------------|---------------------|------------|---------------------|----------| 
| 0 | 0007077a-5d0f-49a4-91d6-0b8dc32837c6 | sxj             | IMPORT       | 2019-12-23 11:33:00 | IMPORT     | 2019-12-23 11:39:24 | 00:06:24 | 
| 1 | 000b03f4-afd8-4371-b79e-2891d5388181 | lfcjx           | UPDATE       | 2019-12-23 14:11:00 | UPDATE     | 2019-12-23 14:11:02 | 00:00:02 | 
| 2 | 000b147d-1d2e-41a6-8ab6-d8888c50e12a | ozraghkiejdns   | APPROVE      | 2019-07-28 09:04:00 | APPROVE    | 2019-07-28 09:06:37 | 00:02:37 | 
| 3 | 002e3d10-33e4-45aa-b882-0d8fab4c5819 | lfcjx           | APPROVE      | 2019-03-18 15:40:00 | APPROVE    | 2019-03-18 15:47:54 | 00:07:54 | 
| 4 | 0031390e-202f-447f-9de0-044851d7e78a | gkxqjitrflwmuay | IMPORT       | 2019-09-27 16:10:00 | IMPORT     | 2019-09-27 16:10:02 | 00:00:02 | 


Having this data frame in place all the following calculations should be much easier. You may also introduce additional attributes like weekday etc. depending on what you're going to compute further. It's common practice to use de-normalization for the sake of performance.

**Comment by sleshJdev** [comment]
Also, you can develop columns num_success, num_warnings and num_failed

**Comment by sleshJdev** [comment]
Check this function https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pivot.html?highlight=pivot#pandas.DataFrame.pivot

It will do what you want

**Comment by aliaksandr-tsukanau** [comment]
Nice and correct solution, well done!

I have an idea how to speed up and simplify the creation of `user_actions`. Currently the problem is that we invoke custom Python functions using `apply` - it is inefficient and hard to read. But we can use optimized vectorized operations that Pandas provides.

1. You can read the data as is and add "action" and "state" columns:
```python
original_data = pd.read_csv('user_actions.csv', 
                            usecols = lambda col : col != "currentUser",
                            parse_dates=['createdDateTime']).set_index('actionId')
original_data['state'] = original_data['message'].str.extract(...)
original_data['action'] = original_data['message'].str.extract(...)
```
Notice that you don't need to call `apply` - it's better to use [pandas.Series.str.extract](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.extract.html)

2. Let's create two separate views for rows with "START" and "END":
```python
start_df = original_data[original_data['state'] == 'START']
end_df = original_data[original_data['state'] == 'END']
```

3. Create empty `user_actions` and use "actionId" from original data as index. But drop duplicates.
```python
user_actions = pd.DataFrame(index=original_data.index.drop_duplicates())
```

4. Fill all the required columns from `start_df` or `end_df`. It will automatically align using actionId index.
```python
user_actions['action'] = start_df['action']
user_actions['started_at'] = start_df['createdDateTime']
user_actions['ended_at'] = end_df['createdDateTime']
user_actions['duration'] = user_actions['ended_at'] - user_actions['started_at']
user_actions['week'] = ... # find a vector function for that
user_actions['num_records'] = start_df['message'].str.extract(...) # write a regex to extract number after "num_records="
user_actions['num_success'] = end_df['message'].str.extract(...)
... # add more columns
```

5. Insert 0 for searches:
```python
user_actions.loc[user_actions['action'] == 'SEARCH', ['num_records', 'num_success', 'num_failed', 'num_warning']] = 0
```

6. Reset index so that the rest of you notebook works as before:
```python
user_actions.reset_index(inplace=True)
```

That's it! No custom functions, no joins, only Pandas beauty.

Feel free to contact me in Skype. We can have a call so that I explain more if needed

**Comment by Tasenga** [comment]
Thanks for such a simple and effective way. I was very surprised that pandas can concatenate data from several tables by index so easily, at first I tried something similar using the join method.

**Comment by aliaksandr-tsukanau** [comment]
Excellent!

I’ve just found out that this line can be much simpler than that:
```
not_srch_act = user_actions.drop(
    user_actions[user_actions['action'].apply(lambda x: x == 'SEARCH')].index)
```
It can be just
```
not_srch_act = user_actions[user_actions[‘action’] != ‘SEARCH’]
```

@sleshJdev looks good to me, please approve

**Comment by Tasenga** [comment]
> 
> 
> Excellent!
> 
> I’ve just found out that this line can be much simpler than that:
> 
> ```
> not_srch_act = user_actions.drop(
>     user_actions[user_actions['action'].apply(lambda x: x == 'SEARCH')].index)
> ```
> 
> It can be just
> 
> ```
> not_srch_act = user_actions[user_actions[‘action’] != ‘SEARCH’]
> ```
> 
> @sleshJdev looks good to me, please approve

Done

**Comment by sleshJdev** [comment]
Looks really nice. 

Just qq @alex-tsukanov could you please explain why it's more efficient. As we discussed, for some operations pandas don't copy the actual data but provide you a "view" instead. It would be great if you provide some links/keywords to pandas documentation. 

Again, thanks for your help.

**Comment by sleshJdev** [comment]
@Tasenga could you please refactor this one as well
```
user_actions[user_actions['action']\
            .apply(lambda x: x == 'SEARCH')][['actionId', 'action', 'duration']]\
            .sort_values(['duration'], ascending=False)[:10]
```

Please, make sure to don't use .apply at all. I guess it's possible when we have nicely prepared data.

**Comment by Tasenga** [comment]
> 
> 
> @Tasenga could you please refactor this one as well
> 
> ```
> user_actions[user_actions['action']\
>             .apply(lambda x: x == 'SEARCH')][['actionId', 'action', 'duration']]\
>             .sort_values(['duration'], ascending=False)[:10]
> ```
> 
> Please, make sure to don't use .apply at all. I guess it's possible when we have nicely prepared data.

Done (there was only this apply).

**Review by sleshJdev — APPROVED** [review]
Well done!

**Comment by aliaksandr-tsukanau** [comment]
@Tasenga 
Я позволю себе перейти на русский, потому что тема довольно продвинутая. Давай обсудим, почему вообще Pandas хвалят за производительную работу с таблицами и почему часто делают акцент на том что это библиотека векторных вычислений, хранящая данные в колоночном формате.

Что касается векторности, здесь легко ввести себя в заблуждение. Допустим, мы векторно умножаем столбец на 2:
```python
df = pd.DataFrame({'original': [1, 2, 3]})
df['doubled'] = df['original'] * 2
```
Можно предположить, будто Pandas будет распараллеливать вычисления, ведь можно делать умножение для любых строк таблицы одновременно. Но это не так, Pandas всё делает последовательно: сколько бы у вас ни было ядер, Pandas использует только одно. Тогда почему код выше намного эффективнее вот этого?
```python
df['doubled'] = df['original'].apply(lambda x: x*2)
```

По большей части потому что Python медленный язык. `apply` - очень гибкая штука и иногда без него никак, но вообще вызывать Python-функцию для каждой строки таблицы - очень дорого. Pandas предоставляет огромный набор векторных операций специально чтобы избежать вычислений на Python и переложить их на C. Но самого C было бы мало для классной скорости, если бы Pandas не хранил данные в колоночном формате. Почему это важно?

Возьмём для наглядности строку из решения:
```python
user_actions['week'] = start_df['createdDateTime'].dt.week
```
Pandas поочерёдно проходит по всем значениям столбца 'createdDateTime' в цикле на C. Когда Pandas обрабатывает текущее значение, он загружает его в кэш процессора. Но данные из оперативки в кэш могут передаваться только блоками фиксированного размера - линиями кэша (cache line), а значит вместе с обрабатываемой ячейкой в кэш попадут и ещё 64 байта данных (для core i7), которые попросту лежали в оперативке рядом. Если бы Pandas хранил данные построчно, соседними оказались бы данные других столбцов, которые в общем-то бесполезны. Однако колоночный формат означает, что все значения столбца 'createdDateTime' - это непрерывный блок памяти, а значит в линии кэша будут как раз следующие значения, необходимые для обработки. Естественно это сильно ускоряет процесс, т.к. позволяет процессору не идти ещё раз в оперативку, а брать следующее значение из кэша, что на несколько порядков быстрее. (Термин для загугливания - CPU cache miss.)

@sleshJdev Что касается вопроса о копировании данных, я не думаю что мы тут делаем что-то особенное. Я ошибочно назвал start_df и end_df вьюхами - нет, это копии данных (и это отлично, ведь все данные становятся подряд в памяти и нет cache misses). У пандаса довольно сложная логика, по которой он решает когда копировать данные, а когда делать ссылку, и можно быть уверенным, что она сделана достаточно оптимально. Тут в принципе единственное важное о чём стоит думать это [chained indexing](http://pandas-docs.github.io/pandas-docs-travis/user_guide/indexing.html#indexing-view-versus-copy).

TL;DR: замена apply на векторные функции очень эффективная оптимизация. Как писали не помню где, If you're not sure whether you should be using apply, you probably shouldn't.

**Comment by Tasenga** [comment]
Спасибо за уделенное проекту время и такие глубокие пояснения функциональности pandas. Теперь это более чем понятно. Отличный кейс для сравнения процессов с точки зрения продуктивности.
